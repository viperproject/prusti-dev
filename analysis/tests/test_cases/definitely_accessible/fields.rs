#![allow(dropping_references)]

#[derive(Clone, Default)]
struct T {
    // Wrap in Box to have non-Copy types
    value1: Box<u32>,
    value2: Box<u32>,
    value3: Box<u32>,
}

#[analyzer::run]
fn main() {
    let mut x = T::default();
    // Move out value1
    drop(x.value1);
    // Block value2
    let borrow_value2 = &mut x.value2;
    let block_value2 = &borrow_value2;
    // Freeze value3
    let shared_ref = &x.value3;
    // Nothing should be owned here
    drop(shared_ref);
    // Now value3 should be owned
    drop(block_value2);
    // Now value2 should be owned
    x.value1 = Box::new(123);
    // Now everything should be owned
    drop(x);
    // Now the state should be empty
}
