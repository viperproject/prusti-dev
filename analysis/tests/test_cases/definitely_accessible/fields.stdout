Analyzing file $DIR/fields.rs using DefinitelyAccessibleAnalysis...
Result for function main():
{
  "bb0": [
    [
      [
        "state:",
        {
          "accessible": [],
          "owned": []
        },
        "statement: StorageLive(_1)"
      ]
    ],
    "state before terminator:",
    {
      "accessible": [],
      "owned": []
    },
    "terminator: _1 = <T as std::default::Default>::default() -> [return: bb1, unwind: bb14]",
    {
      "bb1": [
        "state:",
        {
          "accessible": [
            "_1"
          ],
          "owned": [
            "_1"
          ]
        }
      ],
      "bb14": [
        "state:",
        {
          "accessible": [],
          "owned": []
        }
      ]
    }
  ],
  "bb1": [
    [
      [
        "state:",
        {
          "accessible": [
            "_1"
          ],
          "owned": [
            "_1"
          ]
        },
        "statement: FakeRead(ForLet(None), _1)"
      ],
      [
        "state:",
        {
          "accessible": [
            "_1"
          ],
          "owned": [
            "_1"
          ]
        },
        "statement: StorageLive(_2)"
      ],
      [
        "state:",
        {
          "accessible": [
            "_1"
          ],
          "owned": [
            "_1"
          ]
        },
        "statement: StorageLive(_3)"
      ],
      [
        "state:",
        {
          "accessible": [
            "_1"
          ],
          "owned": [
            "_1"
          ]
        },
        "statement: _3 = move (_1.0: std::boxed::Box<u32>)"
      ]
    ],
    "state before terminator:",
    {
      "accessible": [
        "(_1.1: std::boxed::Box<u32>)",
        "(_1.2: std::boxed::Box<u32>)",
        "_3"
      ],
      "owned": [
        "(_1.1: std::boxed::Box<u32>)",
        "(_1.2: std::boxed::Box<u32>)",
        "_3"
      ]
    },
    "terminator: _2 = std::mem::drop::<std::boxed::Box<u32>>(move _3) -> [return: bb2, unwind: bb12]",
    {
      "bb12": [
        "state:",
        {
          "accessible": [],
          "owned": []
        }
      ],
      "bb2": [
        "state:",
        {
          "accessible": [
            "(_1.1: std::boxed::Box<u32>)",
            "(_1.2: std::boxed::Box<u32>)",
            "_2"
          ],
          "owned": [
            "(_1.1: std::boxed::Box<u32>)",
            "(_1.2: std::boxed::Box<u32>)",
            "_2"
          ]
        }
      ]
    }
  ],
  "bb2": [
    [
      [
        "state:",
        {
          "accessible": [
            "(_1.1: std::boxed::Box<u32>)",
            "(_1.2: std::boxed::Box<u32>)",
            "_2"
          ],
          "owned": [
            "(_1.1: std::boxed::Box<u32>)",
            "(_1.2: std::boxed::Box<u32>)",
            "_2"
          ]
        },
        "statement: StorageDead(_3)"
      ],
      [
        "state:",
        {
          "accessible": [
            "(_1.1: std::boxed::Box<u32>)",
            "(_1.2: std::boxed::Box<u32>)",
            "_2"
          ],
          "owned": [
            "(_1.1: std::boxed::Box<u32>)",
            "(_1.2: std::boxed::Box<u32>)",
            "_2"
          ]
        },
        "statement: StorageDead(_2)"
      ],
      [
        "state:",
        {
          "accessible": [
            "(_1.1: std::boxed::Box<u32>)",
            "(_1.2: std::boxed::Box<u32>)"
          ],
          "owned": [
            "(_1.1: std::boxed::Box<u32>)",
            "(_1.2: std::boxed::Box<u32>)"
          ]
        },
        "statement: StorageLive(_4)"
      ],
      [
        "state:",
        {
          "accessible": [
            "(_1.1: std::boxed::Box<u32>)",
            "(_1.2: std::boxed::Box<u32>)"
          ],
          "owned": [
            "(_1.1: std::boxed::Box<u32>)",
            "(_1.2: std::boxed::Box<u32>)"
          ]
        },
        "statement: _4 = &mut (_1.1: std::boxed::Box<u32>)"
      ],
      [
        "state:",
        {
          "accessible": [
            "(_1.2: std::boxed::Box<u32>)",
            "_4"
          ],
          "owned": [
            "(_1.2: std::boxed::Box<u32>)",
            "_4"
          ]
        },
        "statement: FakeRead(ForLet(None), _4)"
      ],
      [
        "state:",
        {
          "accessible": [
            "(_1.2: std::boxed::Box<u32>)",
            "_4"
          ],
          "owned": [
            "(_1.2: std::boxed::Box<u32>)",
            "_4"
          ]
        },
        "statement: StorageLive(_5)"
      ],
      [
        "state:",
        {
          "accessible": [
            "(_1.2: std::boxed::Box<u32>)",
            "_4"
          ],
          "owned": [
            "(_1.2: std::boxed::Box<u32>)",
            "_4"
          ]
        },
        "statement: _5 = &_4"
      ],
      [
        "state:",
        {
          "accessible": [
            "(_1.2: std::boxed::Box<u32>)",
            "_5"
          ],
          "owned": [
            "(_1.2: std::boxed::Box<u32>)",
            "_5"
          ]
        },
        "statement: FakeRead(ForLet(None), _5)"
      ],
      [
        "state:",
        {
          "accessible": [
            "(_1.2: std::boxed::Box<u32>)",
            "_5"
          ],
          "owned": [
            "(_1.2: std::boxed::Box<u32>)",
            "_5"
          ]
        },
        "statement: StorageLive(_6)"
      ],
      [
        "state:",
        {
          "accessible": [
            "(_1.2: std::boxed::Box<u32>)",
            "_5"
          ],
          "owned": [
            "(_1.2: std::boxed::Box<u32>)",
            "_5"
          ]
        },
        "statement: _6 = &(_1.2: std::boxed::Box<u32>)"
      ],
      [
        "state:",
        {
          "accessible": [
            "(_1.2: std::boxed::Box<u32>)",
            "_5",
            "_6"
          ],
          "owned": [
            "_5",
            "_6"
          ]
        },
        "statement: FakeRead(ForLet(None), _6)"
      ],
      [
        "state:",
        {
          "accessible": [
            "(_1.2: std::boxed::Box<u32>)",
            "_5",
            "_6"
          ],
          "owned": [
            "_5",
            "_6"
          ]
        },
        "statement: StorageLive(_7)"
      ],
      [
        "state:",
        {
          "accessible": [
            "(_1.2: std::boxed::Box<u32>)",
            "_5",
            "_6"
          ],
          "owned": [
            "_5",
            "_6"
          ]
        },
        "statement: StorageLive(_8)"
      ],
      [
        "state:",
        {
          "accessible": [
            "(_1.2: std::boxed::Box<u32>)",
            "_5",
            "_6"
          ],
          "owned": [
            "_5",
            "_6"
          ]
        },
        "statement: _8 = _6"
      ]
    ],
    "state before terminator:",
    {
      "accessible": [
        "(_1.2: std::boxed::Box<u32>)",
        "_5",
        "_8"
      ],
      "owned": [
        "_5",
        "_8"
      ]
    },
    "terminator: _7 = std::mem::drop::<&std::boxed::Box<u32>>(move _8) -> [return: bb3, unwind: bb13]",
    {
      "bb13": [
        "state:",
        {
          "accessible": [],
          "owned": []
        }
      ],
      "bb3": [
        "state:",
        {
          "accessible": [
            "(_1.2: std::boxed::Box<u32>)",
            "_5",
            "_7"
          ],
          "owned": [
            "(_1.2: std::boxed::Box<u32>)",
            "_5",
            "_7"
          ]
        }
      ]
    }
  ],
  "bb3": [
    [
      [
        "state:",
        {
          "accessible": [
            "(_1.2: std::boxed::Box<u32>)",
            "_5",
            "_7"
          ],
          "owned": [
            "(_1.2: std::boxed::Box<u32>)",
            "_5",
            "_7"
          ]
        },
        "statement: StorageDead(_8)"
      ],
      [
        "state:",
        {
          "accessible": [
            "(_1.2: std::boxed::Box<u32>)",
            "_5",
            "_7"
          ],
          "owned": [
            "(_1.2: std::boxed::Box<u32>)",
            "_5",
            "_7"
          ]
        },
        "statement: StorageDead(_7)"
      ],
      [
        "state:",
        {
          "accessible": [
            "(_1.2: std::boxed::Box<u32>)",
            "_5"
          ],
          "owned": [
            "(_1.2: std::boxed::Box<u32>)",
            "_5"
          ]
        },
        "statement: StorageLive(_9)"
      ],
      [
        "state:",
        {
          "accessible": [
            "(_1.2: std::boxed::Box<u32>)",
            "_5"
          ],
          "owned": [
            "(_1.2: std::boxed::Box<u32>)",
            "_5"
          ]
        },
        "statement: StorageLive(_10)"
      ],
      [
        "state:",
        {
          "accessible": [
            "(_1.2: std::boxed::Box<u32>)",
            "_5"
          ],
          "owned": [
            "(_1.2: std::boxed::Box<u32>)",
            "_5"
          ]
        },
        "statement: _10 = _5"
      ]
    ],
    "state before terminator:",
    {
      "accessible": [
        "(_1.2: std::boxed::Box<u32>)",
        "_10"
      ],
      "owned": [
        "(_1.2: std::boxed::Box<u32>)",
        "_10"
      ]
    },
    "terminator: _9 = std::mem::drop::<&&mut std::boxed::Box<u32>>(move _10) -> [return: bb4, unwind: bb13]",
    {
      "bb13": [
        "state:",
        {
          "accessible": [],
          "owned": []
        }
      ],
      "bb4": [
        "state:",
        {
          "accessible": [
            "(_1.1: std::boxed::Box<u32>)",
            "(_1.2: std::boxed::Box<u32>)",
            "_9"
          ],
          "owned": [
            "(_1.1: std::boxed::Box<u32>)",
            "(_1.2: std::boxed::Box<u32>)",
            "_9"
          ]
        }
      ]
    }
  ],
  "bb4": [
    [
      [
        "state:",
        {
          "accessible": [
            "(_1.1: std::boxed::Box<u32>)",
            "(_1.2: std::boxed::Box<u32>)",
            "_9"
          ],
          "owned": [
            "(_1.1: std::boxed::Box<u32>)",
            "(_1.2: std::boxed::Box<u32>)",
            "_9"
          ]
        },
        "statement: StorageDead(_10)"
      ],
      [
        "state:",
        {
          "accessible": [
            "(_1.1: std::boxed::Box<u32>)",
            "(_1.2: std::boxed::Box<u32>)",
            "_9"
          ],
          "owned": [
            "(_1.1: std::boxed::Box<u32>)",
            "(_1.2: std::boxed::Box<u32>)",
            "_9"
          ]
        },
        "statement: StorageDead(_9)"
      ],
      [
        "state:",
        {
          "accessible": [
            "(_1.1: std::boxed::Box<u32>)",
            "(_1.2: std::boxed::Box<u32>)"
          ],
          "owned": [
            "(_1.1: std::boxed::Box<u32>)",
            "(_1.2: std::boxed::Box<u32>)"
          ]
        },
        "statement: StorageLive(_11)"
      ]
    ],
    "state before terminator:",
    {
      "accessible": [
        "(_1.1: std::boxed::Box<u32>)",
        "(_1.2: std::boxed::Box<u32>)"
      ],
      "owned": [
        "(_1.1: std::boxed::Box<u32>)",
        "(_1.2: std::boxed::Box<u32>)"
      ]
    },
    "terminator: _11 = std::boxed::Box::<u32>::new(const 123_u32) -> [return: bb5, unwind: bb13]",
    {
      "bb13": [
        "state:",
        {
          "accessible": [],
          "owned": []
        }
      ],
      "bb5": [
        "state:",
        {
          "accessible": [
            "(_1.1: std::boxed::Box<u32>)",
            "(_1.2: std::boxed::Box<u32>)",
            "_11"
          ],
          "owned": [
            "(_1.1: std::boxed::Box<u32>)",
            "(_1.2: std::boxed::Box<u32>)",
            "_11"
          ]
        }
      ]
    }
  ],
  "bb5": [
    [],
    "state before terminator:",
    {
      "accessible": [
        "(_1.1: std::boxed::Box<u32>)",
        "(_1.2: std::boxed::Box<u32>)",
        "_11"
      ],
      "owned": [
        "(_1.1: std::boxed::Box<u32>)",
        "(_1.2: std::boxed::Box<u32>)",
        "_11"
      ]
    },
    "terminator: drop((_1.0: std::boxed::Box<u32>)) -> [return: bb6, unwind: bb7]",
    {
      "bb6": [
        "state:",
        {
          "accessible": [
            "(_1.1: std::boxed::Box<u32>)",
            "(_1.2: std::boxed::Box<u32>)",
            "_11"
          ],
          "owned": [
            "(_1.1: std::boxed::Box<u32>)",
            "(_1.2: std::boxed::Box<u32>)",
            "_11"
          ]
        }
      ],
      "bb7": [
        "state:",
        {
          "accessible": [],
          "owned": []
        }
      ]
    }
  ],
  "bb6": [
    [
      [
        "state:",
        {
          "accessible": [
            "(_1.1: std::boxed::Box<u32>)",
            "(_1.2: std::boxed::Box<u32>)",
            "_11"
          ],
          "owned": [
            "(_1.1: std::boxed::Box<u32>)",
            "(_1.2: std::boxed::Box<u32>)",
            "_11"
          ]
        },
        "statement: (_1.0: std::boxed::Box<u32>) = move _11"
      ]
    ],
    "state before terminator:",
    {
      "accessible": [
        "_1"
      ],
      "owned": [
        "_1"
      ]
    },
    "terminator: drop(_11) -> [return: bb8, unwind: bb13]",
    {
      "bb13": [
        "state:",
        {
          "accessible": [],
          "owned": []
        }
      ],
      "bb8": [
        "state:",
        {
          "accessible": [
            "_1"
          ],
          "owned": [
            "_1"
          ]
        }
      ]
    }
  ],
  "bb7": [
    [
      [
        "state:",
        {
          "accessible": [],
          "owned": []
        },
        "statement: (_1.0: std::boxed::Box<u32>) = move _11"
      ]
    ],
    "state before terminator:",
    {
      "accessible": [
        "(_1.0: std::boxed::Box<u32>)"
      ],
      "owned": [
        "(_1.0: std::boxed::Box<u32>)"
      ]
    },
    "terminator: drop(_11) -> [return: bb13, unwind terminate]",
    {
      "bb13": [
        "state:",
        {
          "accessible": [
            "(_1.0: std::boxed::Box<u32>)"
          ],
          "owned": [
            "(_1.0: std::boxed::Box<u32>)"
          ]
        }
      ]
    }
  ],
  "bb8": [
    [
      [
        "state:",
        {
          "accessible": [
            "_1"
          ],
          "owned": [
            "_1"
          ]
        },
        "statement: StorageDead(_11)"
      ],
      [
        "state:",
        {
          "accessible": [
            "_1"
          ],
          "owned": [
            "_1"
          ]
        },
        "statement: StorageLive(_12)"
      ],
      [
        "state:",
        {
          "accessible": [
            "_1"
          ],
          "owned": [
            "_1"
          ]
        },
        "statement: StorageLive(_13)"
      ],
      [
        "state:",
        {
          "accessible": [
            "_1"
          ],
          "owned": [
            "_1"
          ]
        },
        "statement: _13 = move _1"
      ]
    ],
    "state before terminator:",
    {
      "accessible": [
        "_13"
      ],
      "owned": [
        "_13"
      ]
    },
    "terminator: _12 = std::mem::drop::<T>(move _13) -> [return: bb9, unwind: bb11]",
    {
      "bb11": [
        "state:",
        {
          "accessible": [],
          "owned": []
        }
      ],
      "bb9": [
        "state:",
        {
          "accessible": [
            "_12"
          ],
          "owned": [
            "_12"
          ]
        }
      ]
    }
  ],
  "bb9": [
    [
      [
        "state:",
        {
          "accessible": [
            "_12"
          ],
          "owned": [
            "_12"
          ]
        },
        "statement: StorageDead(_13)"
      ],
      [
        "state:",
        {
          "accessible": [
            "_12"
          ],
          "owned": [
            "_12"
          ]
        },
        "statement: StorageDead(_12)"
      ],
      [
        "state:",
        {
          "accessible": [],
          "owned": []
        },
        "statement: _0 = const ()"
      ],
      [
        "state:",
        {
          "accessible": [
            "_0"
          ],
          "owned": [
            "_0"
          ]
        },
        "statement: StorageDead(_6)"
      ],
      [
        "state:",
        {
          "accessible": [
            "_0"
          ],
          "owned": [
            "_0"
          ]
        },
        "statement: StorageDead(_5)"
      ],
      [
        "state:",
        {
          "accessible": [
            "_0"
          ],
          "owned": [
            "_0"
          ]
        },
        "statement: StorageDead(_4)"
      ]
    ],
    "state before terminator:",
    {
      "accessible": [
        "_0"
      ],
      "owned": [
        "_0"
      ]
    },
    "terminator: drop(_1) -> [return: bb10, unwind: bb14]",
    {
      "bb10": [
        "state:",
        {
          "accessible": [
            "_0"
          ],
          "owned": [
            "_0"
          ]
        }
      ],
      "bb14": [
        "state:",
        {
          "accessible": [],
          "owned": []
        }
      ]
    }
  ],
  "bb10": [
    [
      [
        "state:",
        {
          "accessible": [
            "_0"
          ],
          "owned": [
            "_0"
          ]
        },
        "statement: StorageDead(_1)"
      ]
    ],
    "state before terminator:",
    {
      "accessible": [
        "_0"
      ],
      "owned": [
        "_0"
      ]
    },
    "terminator: return",
    {}
  ],
  "bb11": [
    [],
    "state before terminator:",
    {
      "accessible": [],
      "owned": []
    },
    "terminator: drop(_13) -> [return: bb13, unwind terminate]",
    {
      "bb13": [
        "state:",
        {
          "accessible": [],
          "owned": []
        }
      ]
    }
  ],
  "bb12": [
    [],
    "state before terminator:",
    {
      "accessible": [],
      "owned": []
    },
    "terminator: drop(_3) -> [return: bb13, unwind terminate]",
    {
      "bb13": [
        "state:",
        {
          "accessible": [],
          "owned": []
        }
      ]
    }
  ],
  "bb13": [
    [],
    "state before terminator:",
    {
      "accessible": [],
      "owned": []
    },
    "terminator: drop(_1) -> [return: bb14, unwind terminate]",
    {
      "bb14": [
        "state:",
        {
          "accessible": [],
          "owned": []
        }
      ]
    }
  ],
  "bb14": [
    [],
    "state before terminator:",
    {
      "accessible": [],
      "owned": []
    },
    "terminator: resume",
    {}
  ]
}
